pallets/ethereum-transaction/src/lib.rs

```diff
@@ -0,0 +1,206 @@
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+#![cfg_attr(not(feature = "std"), no_std)]
+
+use cfg_traits::ethereum::EthereumTransactor;
+use ethereum::{LegacyTransaction, TransactionAction, TransactionSignature, TransactionV2};
+use fp_evm::CallOrCreateInfo;
+use frame_support::{
+	dispatch::{DispatchErrorWithPostInfo, PostDispatchInfo},
+	pallet_prelude::*,
+};
+pub use pallet::*;
+use pallet_evm::{ExitError, ExitFatal, ExitReason};
+use sp_core::{H160, H256, U256};
+use sp_std::vec::Vec;
+
+#[cfg(test)]
+mod mock;
+
+#[cfg(test)]
+mod tests;
+
+const TRANSACTION_RECOVERY_ID: u64 = 42;
+
+#[frame_support::pallet]
+pub mod pallet {
+	use super::*;
+
+	#[pallet::pallet]
+	#[pallet::generate_store(pub (super) trait Store)]
+	pub struct Pallet<T>(_);
+
+	#[pallet::config]
+	pub trait Config: frame_system::Config + pallet_ethereum::Config {
+		/// The event type.
+		type RuntimeEvent: From<Event<Self>> + IsType<<Self as frame_system::Config>::RuntimeEvent>;
+	}
+
+	/// Storage for nonce.
+	#[pallet::storage]
+	pub(crate) type Nonce<T: Config> = StorageValue<_, U256, ValueQuery>;
+
+	#[pallet::event]
+	#[pallet::generate_deposit(pub (super) fn deposit_event)]
+	pub enum Event<T: Config> {
+		/// A call was executed.
+		Executed {
+			from: H160,
+			to: H160,
+			exit_reason: ExitReason,
+			value: Vec<u8>,
+		},
+	}
+
+	impl<T: Config> Pallet<T> {
+		fn get_transaction_signature() -> Option<TransactionSignature> {
+			//TODO(cdamian): Same signature as the one in ethereum-xcm.
+			TransactionSignature::new(
+				TRANSACTION_RECOVERY_ID,
+				H256::from_low_u64_be(1u64),
+				H256::from_low_u64_be(1u64),
+			)
+		}
+	}
+
+	impl<T: Config> EthereumTransactor for Pallet<T> {
+		/// This implementation serves as a wrapper around the Ethereum pallet
+		/// execute functionality. It keeps track of the nonce used for each
+		/// call and builds a fake signature for executing the provided call.
+		///
+		/// NOTE - The execution fees are charged by the Ethereum pallet,
+		/// we only have to charge for the nonce read operation.
+		fn call(
+			from: H160,
+			to: H160,
+			data: &[u8],
+			value: U256,
+			gas_price: U256,
+			gas_limit: U256,
+		) -> DispatchResultWithPostInfo {
+			let nonce = Nonce::<T>::get();
+			let read_weight = T::DbWeight::get().reads(1);
+
+			let signature =
+				Pallet::<T>::get_transaction_signature().ok_or(DispatchErrorWithPostInfo {
+					post_info: PostDispatchInfo {
+						actual_weight: Some(read_weight),
+						pays_fee: Pays::Yes,
+					},
+					error: DispatchError::Other("Failed to create transaction signature"),
+				})?;
+
+			let transaction = TransactionV2::Legacy(LegacyTransaction {
+				nonce,
+				gas_price,
+				gas_limit,
+				action: TransactionAction::Call(to),
+				value,
+				input: data.into(),
+				signature,
+			});
+
+			Nonce::<T>::put(nonce.saturating_add(U256::one()));
```

Comment:
I know that the maximum number is astronomically large, but I would feel safer to still throw if it fails to increment

```suggestion
			Nonce::<T>::put(nonce.ensure_add(U256::one())?);
```

Reply:
OK, will change.

Reply:
It seems that the `EnsureAdd` trait is not satisfied for this type. Am I missing something?

Reply:
You need to import `sp_runtime::traits::EnsureAdd`

Reply:
I did that, I still get some errors regarding missing trait bounds. I'm confused.

Reply:
Are we OK with leaving it as it is then?

Reply:
From my POV, I think it's ok saturating here. There are more probabilities of having a hash collision, than having an overflow in a `U256` adding ones.

Reply:
Yeah, leave as is!